### PR TITLE
"make world" and dune default target don't build coqide

### LINKED ISSUE #10821: FR: build coq using dune without coqide dependencies
#### Description of the problem

When trying to build coq I get the following error:

```
[build@archlinux coq]$ make -f Makefile.dune world
dune build  @vodeps
dune exec ./tools/coq_dune.exe _build/default/.vfiles.d
dune build  @install
File "ide/dune", line 49, characters 12-22:
49 |  (libraries coqide_gui))
                 ^^^^^^^^^^
Error: Library "coqide_gui" in _build/default/ide is hidden (optional with
unavailable dependencies).
Hint: try: dune external-lib-deps --missing @install
make: *** [Makefile.dune:55: world] Error 1
```

in the following environment:

```yaml
image: archlinux
packages:
- ocaml
- ocaml-num
- camlp5
- dune
- time
```

This is probably because I do not have gtk installed, and `make -f Makefile.dune world` expects it is. @JasonGross  says it would be great to have a CI target that tests building coq when gtk and such are not installed.

I am currently just ignoring that error in my build script.

cc @ejgallego 

#### Coq Version

efed96f90c73d8e33a6a1eb0f4161dbdd7f7f533

### PR DESCRIPTION
**Pull request by SkySkimmer**
makefile add coqide target

The current behaviour seems mostly annoying to those who don't care about coqide.

Closes #10821

### CODE REVIEW

File: Makefile
Comment (silene) [inline Makefile:149]: Would people mind if we were to just remove `coqide` from `NONDOC_INSTALL_TARGETS` and add a dedicated rule for it?
```make
coqide:
        dune build $(DUNEOPT) coqide.install
```
(As a pure CoqIDE user, I certainly would not mind.)

File: Makefile
Comment (silene) [inline Makefile:153]: Did you remove `dunestrap` on purpose?

**Review by ejgallego — COMMENTED** [review]
Fine to me, however note that we cannot make everyone happy.

Old setup was like this, then we got the complaint that people wrote a patch, did `make world`, things went fine, then build failed in CI as they forgot to update CoqIDE stuff...

So YMMV. In general I have found non-deterministic build target to often more of a pain than a gain.

A `world-noide` target may be more deterministic, and we keep `make world` for "really build everything in the tree"

File: Makefile
Comment (SkySkimmer) [reply Makefile:153]: oops no

**Comment by ejgallego** [comment]
> The current behaviour seems mostly annoying to those who don't care about coqide.

Who are those that don't care about coqide?

**Comment by SkySkimmer** [comment]
A few people at the cudw had dune always complaining about missing lablgtk

**Comment by ejgallego** [comment]
> A few people at the cudw had dune always complaining about missing lablgtk

Why they can't install `lablgtk` instead?

The more I look into this the more I like that I cannot know what `make world` does anymore without having to request people to list their whole opam package setting.

**Comment by andres-erbsen** [comment]
I also find the current behavior annoying. I would be satisfied with an everything-but-coqide target under any name (I already carry one locally). Automatically skipping coqide in `make world` when dependencies are not present would be fine if it generated a warning, but having unknowing users get different builds from the same command seems undesirable indeed. Separately,  a "local ci" target would be very valuable too, but `make world` is not that -- it does not check the doc or whitespace, for example, so I think this consideration isn't pressing here.

**Comment by ejgallego** [comment]
> Separately, a "local ci" target would be very valuable too, but `make world` is not that -- it does not check the doc or whitespace, for example, so I think this consideration isn't pressing here.

`make world` has a clear meaning from a long time and is "build all code in the repository"

Indeed, it only covers compilation, and not tests or documentation, because that is historically done in other targets, and usually not needed to submit a PR.

However, a failing `make world` for CoqIDE does mean CI wasted time.

**Comment by ejgallego** [comment]
> I would be satisfied with an everything-but-coqide target under any name (I already carry one locally).

I think that would be the best solution, `make world-noide` or something like that.

> Automatically skipping coqide in `make world` when dependencies are not present would be fine if it generated a warning, but having unknowing users get different builds from the same command seems undesirable indeed.

It generates a warning now, but indeed this goes against the whole philosophy of this PR: either it is OK to build without CoqIDE (and then a warning should not happen), or otherwise users should install lablgtk

**Comment by andres-erbsen** [comment]
Ah, as it already generates a warning, this PR seems like a strict improvement to me -- a late, low-level error message replaced with an immediate, intentionally worded one. People will use `make world` to build without coqide regardless.

**Comment by ejgallego** [comment]
> Ah, as it already generates a warning, this PR seems like a strict improvement to me -- a late, low-level error message replaced with an immediate, intentionally worded one. 

It is not an improvement because previously we didn't succeed, now we do. For example our gitlab setup will need adjustment, as it uses the `world` target.

> People will use `make world` to build without coqide regardless.

They should stop doing that.

**Review by ejgallego — DISMISSED** [review]
CI setup would need changes.

Anyways this seems a bad idea to me, for the reasons given in the discussion.

If people want a different target, let's add a different target, instead of weakening the semantics of `world` and adding maintenance burden to Coq's code.

**Comment by ejgallego** [comment]
>  a late, low-level error message replaced with an immediate, intentionally worded one.

I agree that dune message should be better, I guess there is a bug in Dune already opened.

**Comment by andres-erbsen** [comment]
I don't understand how this would affect the CI, which has lablgtk installed.

> They should stop doing that.

Perhaps this is the core disagreement here, and maybe even more broadly. It is possible that they should, but the point is, we can't make them. Opinionated workflow designs have to compete with the less systematic alternatives and win on their merit, as judged by reveled preferences of the users. We won't get far by blaming others for doing the straightforward thing to get work done. This PR fixes a clear pain point in the developer experience, and I support it.

**Comment by proux01** [comment]
It seems there is a clear majority in favor of the change. Let's just ensure that CI is not broken (apparently it's just a matter of updating `.build-template:base:dev` in .gitlab-ci.yml which seems to be the only place where `make world` is called in the CI) and go ahead.

**Comment by ejgallego** [comment]
> I don't understand how this would affect the CI, which has lablgtk installed.

CI does `make world`, now if for some reason we do a mistake in some other part of the CI setup, we will stop building coqide and we won't realize.

**Comment by ejgallego** [comment]
> It seems there is a clear majority in favor of the change. 

Where is the majority? So far I see only Anders making an opinion and they stating the `world-noide` solution would also work for them.

Anyways, I am not sure "having a clear majority" should be the primary criterion for technical review, but instead we should stick to technical criteria to determine what is best.

So far it is not clear how this proposal solves the problems users were facing, for the reasons stated above.

What is the point of allowing `make world` to succeed with a warning? That means it is still not usable for the people that didn't care about CoqIDE, and introduces other problems to solve (such as more code to maintain + updates to the CI setup)

Having a `world-noide` (or maybe `stdlib` would be a better / more faithful target name) seems to have much better properties.

> Let's just ensure that CI is not broken (apparently it's just a matter of updating `.build-template:base:dev` in .gitlab-ci.yml which seems to be the only place where `make world` is called in the CI) and go ahead.

How do you ensure that?

**Comment by ejgallego** [comment]
> This PR fixes a clear pain point in the developer experience, and I support it.

I don't see how allowing the compilation to go ahead with a warning is a fix for the developer experience.

**Comment by SkySkimmer** [comment]
>However, a failing make world for CoqIDE does mean CI wasted time.

The vast majority of changes don't break coqide

**Comment by ejgallego** [comment]
Things that are still not clear IMHO:

- why the affected users cannot install lablgtk ?
- what is the target audience of "compiling Coq without CoqIDE" ?

  I have a hard time seeing who they are, because as of today, any person trying to contribute to Coq, needs to ensure their code compiles with CoqIDE too.

I am happy if we decide to drop this last requirement, but it seems that decision should be taken in a different PR, and thus move the compilation of CoqIDE to a difference CI job.

**Comment by ejgallego** [comment]
> > However, a failing make world for CoqIDE does mean CI wasted time.
> 
> The vast majority of changes don't break coqide

Sure, but it is still tested in the default build, so I don't see what we gain from this asymmetry.

**Comment by Janno** [comment]
I do not personally use `make world` but I do generally suffer from the lablgtk problem when running `dune build` (without a specific target; maybe that's the same as `make world`?) and also when using the ocaml debugger (which @SkySkimmer taught me how to use during CUDW). For `dune build` I just got used to ignoring the failure. For the debugger, I keep around a stashed change to `dev/dune` that comments out the line mentioning coqide. Of course I keep forgetting to remove the change before I commit and then I also keep forgetting to re-apply that stash after I undo it.

> why the affected users cannot install lablgtk ?

My opam switches are generally as small as possible because I find interacting with opam to be extremely painful and time consuming. So if there is a package I can avoid I will definitely prefer that. Still, I was annoyed enough during CUDW and tried to install lablgtk. Putting aside the fact that `opam install lablgtk` took almost five minutes(!) to come back with a reply, it also wanted me to re-install the entire switch because of nebulous "system changes". This happens 50% of the times that I want install a package. I suspect, but cannot say for certain, that installing the lablgtk package might make this problem even worse because there is now an extra thing tied to my host system that can "change" and force me to spend almost an hour recompiling my entire switch.

I ended up not installing the package during CUDW for fear of destroying my switch (which does happen very regularly when I use opam but maybe that is not opam's fault) and also because it would take too long. But I did install it shortly after the workshop because I was doing so much OCaml debugging work that the stashing really got in the way. As predicted, I lost about an hour waiting for everything to recompile. I have not dared to ask opam to do anything else for me since then but I am sure it will ask me to recompile everything again very soon.

My point is that opam is more of a torture machine than a package manager so installing packages is just generally something that I avoid. It really has nothing to do with lablgtk. That just happens to be one package that does not get installed by default in the various ways that I create switches. And then I am usually too annoyed of opam to install it after the fact.

> what is the target audience of "compiling Coq without CoqIDE" ?
> I have a hard time seeing who they are, because as of today, any person trying to contribute to Coq, needs to ensure their code compiles with CoqIDE too.

I must admit that I have never tested my Coq patches with CoqIDE.

**Comment by andres-erbsen** [comment]
Thank you for writing this up, Janno. My experience is very similar.

**Comment by ejgallego** [comment]
> maybe that's the same as `make world`?

It is not, but close, and if we update the default world target to be non-ide it should be updated too. That's the current setup for `dune build`:

```lisp
(alias
 (name default)
 (deps coq-core.install coq-stdlib.install coqide-server.install coqide.install))
```

c.f.: #14559

**Comment by Janno** [comment]
> @Janno indeed this patch will do nothing to solve your issues with dune build and the debugger; they need to be addressed in a different way.

Indeed. I forgot to add a disclaimer to my comment: it is basically off-topic as far as the actual PR here is concerned and I am sorry to derail the discussion. I have not actually looked at the changes. I just wanted to answer some of your questions and explain why I am generally in favor of being able to build all the parts of coq that I need without having to have extra packages installed for the stuff I don't need.

> Rest of your mail highlights a problem that is core to Coq: external deps do cause pain to developers often, on the other hand, it is just not possible to develop modern software without this. I see large project managing their external deps successfully with either Nix and opam, so if they do, it should be possible for Coq to do so too, but again, that is IMO a discussion that belongs to its own issue (c.f. https://github.com/coq/coq/issues/17169)

I don't think this really is an issue of external dependencies in Coq. In fact, I think Coq could benefit from using more external libraries. It just happens to be the case that I never have lablgtk installed and and due to my past and present experiences with opam I would like to keep it that way if I can get away with it. If coq soon requires more external dependencies they will end up in my switches just from trying to install coq. But as it stands right now, my switches have coq installed but not lablgtk. So the experience is that I have a fully functioning installation of Coq but I cannot develop Coq without annoying error messages. In some sense my problem would be solved by making lablgtk a mandatory dependency of the coq package but there are very good reasons to not do that.

**Comment by proux01** [comment]
> what is the target audience of "compiling Coq without CoqIDE" ?

I feel this includes the vast majority of Coq developers, who often modify Coq but rarely coqide. And as already pointed, the vast majority of those changes don't break coqide. So just not bothering about coqide and just having the CI safeguarding against the rare cases of accidental breakages would seem a very reasonnable solution.

[off topic]
> [...] I find interacting with opam to be extremely painful [...]

If you are wiling to trade compilation time for bandwidth / disk space, Nix + Cachix may be a good alternative.
[/off topic]

> I must admit that I have never tested my Coq patches with CoqIDE.

So do I. I almost never compile it, the Ci checks that I don't break it by making absolutely unrelevant changes and everything is fine.

> We have experience with non-deterministic build targets and they always come to bit people back.
[...] a deterministic, non IDE target, [...] to make the default at some point (which is fine to me).

So maybe the best solution is simply to have `world` (or `all` or `coq` or whatever short name) uncoditionally not build coqide (after all, as you mention it's already the case with many things like doc and tests) and have a separate coqide target. That would also better reflect the OPAM packages since coq doesn't contain coqide.

**Comment by ejgallego** [comment]
> So maybe the best solution is simply to have `world` (or `all` or `coq` or whatever short name) uncoditionally not build coqide (after all, as you mention it's already the case with many things like doc and tests) and have a separate coqide target. That would also better reflect the OPAM packages since coq doesn't contain coqide.

IMHO that's the best solution too, let's make `world` build not coqide, and make a coqide target as @silene suggested.

We can adapt CI in this case easily (build both targets), and eventually split CoqIDE to its own CI job.

Only issue with updating `world` is that we need to advertise this change properly to devs, this is a very common target and thus devs that use it should be aware of the change; I'm not sure we have such a channel actually.

Adding a `world-noide` and `coqide` has the advantage of preserving the meaning of `world`, but if we want `dune build` to match that it doesn't matter.

**Comment by ejgallego** [comment]
Note also https://github.com/coq/coq/pull/16505 , before this, we had `(optional)` in coqide as to try to handle the case of missing lablgtk better.

But as you can see this stuff is subtle, and in general I've found that different developers in the project have different preferences.

(Myself I tend to choose whatever makes the setup simpler / easier to test)

**Comment by proux01** [comment]
> Only issue with updating `world` is that we need to advertise this change properly to devs, this is a very common target and thus devs that use it should be aware of the change; I'm not sure we have such a channel actually.

A changelog entry + a message on Zulip dev should do a good enough job.

**Comment by proux01** [comment]
@SkySkimmer we probably also need to add `make coqide` after https://github.com/coq/coq/blob/aaa0c139a1a117fca1b8a47508acd1ae01ecc2cd/.gitlab-ci.yml#L110

File: .gitlab-ci.yml
Comment (SkySkimmer) [inline .gitlab-ci.yml:113]: I did this instead of changing the script

**Comment by SkySkimmer** [comment]
@coqbot run full ci

File: default.nix
Comment (SkySkimmer) [inline default.nix:114]: @coq/nix-maintainers should coqide be optional in the dune install call at line 123? (regardless of this PR)

Comment (proux01) [reply default.nix:114]: I'm not using that Nix config myself but I'd say yes.

Comment (Zimmi48) [reply default.nix:114]: Indeed, it should. You can do so like this (IINM):

```nix
  installPhase = ''
    runHook preInstall
    dune install --prefix $out --libdir $OCAMLFIND_DESTDIR coq-core coq-stdlib coq coqide-server ${optional buildIde "coqide"}
    runHook postInstall
  '';
```

**Review by proux01 — APPROVED** [review]

**Comment by proux01** [comment]
Will merge once CI is green

**Comment by proux01** [comment]
@coqbot merge now

**Pull request merged**
